### PR TITLE
[G2M] source GitHub releases

### PR DESCRIPTION
**Pull request by woozyking**
releases

(example from monthly custom range, order various slightly for weekly and daily updates, formats remain the same)

### CODE REVIEW

File: sources/github/api.js [hunk 1]
```diff
@@ -85,3 +85,22 @@ module.exports.getPRsCommits = ({ prs, start, end }) => Promise.all(prs.map(
...
   }).map((r) => ({ ...r, pull_request_url: pr.pull_request_url, pr_html_url: `${pr.html_url}/commits/${r.sha}` }))),
 )).then((data) => data.flat())
+
+module.exports.getReleases = async ({ repos, start, end }) => {
+  const urls = repos.reduce((acc, curr) => {
+    if (acc.indexOf(curr.url) < 0) {
+      acc.push(curr.url)
+    }
+    return acc
+  }, [])
```
Comment (vxsl) [inline sources/github/api.js:95]: what about something like `[... new Set(repos.map(({url}) => url))]`?

File: sources/github/api.js [hunk 2]
```diff
@@ -85,3 +85,22 @@ module.exports.getPRsCommits = ({ prs, start, end }) => Promise.all(prs.map(
...
+  const releases = await Promise.all(urls.map((v => client.request({ url: `${v}/releases` }))))
+  const _start = DateTime.fromISO(start, { zone: 'UTC' }).startOf('day')
+  const _end = DateTime.fromISO(end, { zone: 'UTC' }).startOf('day')
+  return releases
+    .map(({ data }) => data)
+    .flat() // this would auto filter empty data arrays
```
Comment (vxsl) [inline sources/github/api.js:101]: this is awesome!

File: sources/github/api.js [hunk 3]
```diff
@@ -85,3 +85,22 @@ module.exports.getPRsCommits = ({ prs, start, end }) => Promise.all(prs.map(
...
 )).then((data) => data.flat())
+
+module.exports.getReleases = async ({ repos, start, end }) => {
+  const urls = repos.reduce((acc, curr) => {
+    if (acc.indexOf(curr.url) < 0) {
+      acc.push(curr.url)
+    }
+    return acc
+  }, [])
+  const releases = await Promise.all(urls.map((v => client.request({ url: `${v}/releases` }))))
```
Comment (vxsl) [inline sources/github/api.js:96]: redundant brackets inside `map`

File: sources/github/index.js [hunk 4]
```diff
@@ -209,3 +209,25 @@ module.exports.formatPreviously = ({ repos, issues, prs, start, end }) => {
 
   return { content, title: `Previously - ${formatDates({ start, end })}` }
 }
+
+module.exports.formatReleases = ({ post, releases, pre = true }) => {
+  if (!releases || releases.length === 0) {
+    return post
+  }
+  let content = `${releases.length} Releases\n`
+  const byProjects = releases.reduce(groupByProj, {})
```
Comment (vxsl) [inline sources/github/index.js:218]: what is the reason for passing `groupByProp` this as a prop instead of just writing it inline here?

**Review by vxsl — APPROVED** [review]
lgtm, just nitpicking

File: sources/github/api.js (hunk 1 above)
Comment (woozyking) [reply sources/github/api.js:95]: good call, this was pretty much copy/pasta from similar functions within the repo (by copilot lol)

File: sources/github/index.js (hunk 4 above)
Comment (woozyking) [reply sources/github/index.js:218]: so i know the intent more clearly when i fully refactor underlying data processing using something like pandas for python -- it's too hacky right now

**Review by hyx131 — APPROVED** [review]

**Review by sallkall — APPROVED** [review]

**Pull request merged**